### ISSUE TITLE
Confusion with alternatives for git-submodule vs git-subtree

### ISSUE THREAD

**Issue by andry81** [ingydotnet/git-subrepo#565]
Have read the readme and this:

https://github.com/ingydotnet/git-subrepo/issues/558

```
git-subrepo - Git Submodule Alternative
```

You should state in the readme that the `git-subrepo` is an alternative to `git-subtree` in first place, because both does merge git worktrees. Git Submodule works as an opposingly different, through the git links as I noted correctly.

I've tried to read the readme and hard to say from the beginning of the text what the extension actually does. Only scrolling to the commands part then, there is said:

>This is similar in feel to git clone. You just specify the remote repo url, and optionally a sub-directory and/or branch name. The repo will be fetched and **merged into the subdir**.

Which is the opposite what the Git Submodule does. And saying in terms of alternative the [vcstool](https://github.com/dirk-thomas/vcstool) is more "alternative" to `git-submodule`.

So I suggest to just fix the confusion.

**Comment by frank-dspeed** [comment]
@andry81 there is no confusion this does what git subtree should do and it does not do it.

This does not only Create a folder and mergs it into that it also adds a tracking branch and other meta to easy move that subtree or lets say folder into its own repo again. 

you can see this as none intrusive MonoRepo Alternativ that is compatible to single repos.

**Comment by andry81** [comment]
I still not get the difference from that explanation versus the `git-subtree`. What means the "can be split out" while it is already merged?

**Comment by frank-dspeed** [comment]
@andry81 you are correct i give you a more understand able example where i use git-subtree i am maintainer of a Own Operating System so a many 100gb Gitrepo this is the full distribution it contains only the source for it. in that repo are folders lets say there is a folder called "3rd_party" in that folder are lets say products from other vendors stuff like chromium and all its subtrees stuff like android and all its subtrees

googlegit by the way also supports subtrees out of the box if your not a googler you can not know that.

Anyway lets go more far when i now need to modify anything in the 3rd_party subtrees subtrees to build my whole 100gb stack and this change turns out to be usefull in the real source project that is cloned as subtree of subtress in 3rd_party

i can split that out lets fix for example Aurora The Window Manager of Chromium it exists as own repo while it is also a subtree of Chromium, LaCrOS, Android, Fuchsia OS, All got a Own Subtree because all need to be able to test and apply indipendent patches to aurora while they all need to be able to contribute back to Aurora it self if it makes sense.

## how does subtree do that?

when you do subtree clone 3rd_party/aurora git://...... it will create not only that folder in your current branch it will create a own branch only containing aurora and its changes this branch is 1:1 with the real source indipendent and can be indipendent be split out again into a single repo 

git subrepo allows you to take a single indipendent repo as subtree dependency in a big mono repo while handling it as it would be a single repo.

**Comment by andry81** [comment]
@frank-dspeed So, are you saying you just use the same approach as in the `git-subtree` plus a branch allocation to be able to push changes back into being merged dependent repository through a branch? If so then, it still closer as an alternative to the `git-subtree` than to the `git-submodule` because in the `git-submodule` a repository does not need to be merged at all.

**Comment by frank-dspeed** [comment]
@andry81 the .gitrepo file is not tracked inside the branch that reflects the 1:1 copy of the imported subrepo so .gitrepo is not really altered to the original repo. as that gets not synced with the 1:1 branch

lets assume aurora again

- main - my-big-project/subrepos/aurora/.gitrepo is not part of
- gitsubrepo/aurora - the 1:1 tracking branch it does not contain .gitrepo

the gitrepo file contains the refenrences to our internal 1:1 tracking branch that we use if we want to split out changes

and your correct all this is needed to reach a higher goal the higher goal is that in my-big project i do not want to have the whole individual commits while i want to preserve individual commits that are done in my big project.

This allows better faster code review for the reviewers of my big project we also need to merge the changes in as they are in fact part of our code as our code will not work without that code we directly merge it we do not want additional downloads in case of emergencys or corruption of the source 3rd party we need to be able to recover anyway. we see that from a Professional Corperated Perspective and not from a individual developer perspective.

hope that makes sense.

**Comment by admorgan** [comment]
@andry81, coming from an svn background before utilizing git I agree that I have occasionally longed for svn externals.

git subrepo is designed to allow you to have a monorepo (a fully self contained repository of code) that has the ability to share sections of that code with other projects.

It will not make a good replacement for svn externals. It inherently has different goals. It on the other hand frequently works better than svn externals because you can never get in a situation where code is missing because a server is down. It also allows you to state why you are using a specific version instead of just having a link.

In my SVN days I ran into significant issues with externals that automatically tracked the head of a branch, that is obviously impossible with subrepo.

You created a list of "features" you are looking for in a git extension. I am curious which of these are essential and of the optional ones which are most important.

**Comment by andry81** [comment]
> You created a list of "features" you are looking for in a git extension. I am curious which of these are essential and of the optional ones which are most important.

I think these must be enough:

1. subtree merge with sparse checkout (partial tree checkout)
2. single list for all externals
3. external link to track head or branch tip

But in case of git worktree merge (which does git-subree and git-subrepo), the track head is just not the case and it is ok.

**Comment by admorgan** [comment]
1. subtree merge with sparse checkout (partial tree checkout).
This is very difficult. There have been a few attempts to do this in the past, but none of them have been successful. The last proposed attempt was to contain a list of files or directories to include in the .gitrepo, then rewrite the history of the external repo to remove everything else when playing out a pull or push. The computations expense of this is very high. He failure rate when things are added/subtracted or moved in the external repo is also exceptionally high. My suggestion is if there is a subsection of a repo you would like to utilize use git subrepo init to break that section out into it's own repo and then use recursive subrepos to maintain this infrastructure.

2. Single list for all externals
Why is this a necessary feature? git subrepo operation on one or more subrepo via each command using the --all. Why is having a single location for all external settings a feature? I personally feel it is an anti-feature as moving data between branching makes a single location of metadata fragile.

3. external link to track head or branch tip
I read your description of testing changes before merging them into the primary worktree, but this goes against what I understood from reading the external link to track branch tips. In svn's external that is following a branch tip it will automatically integrate the newest code. This has caused me a lot of problems on projects that were utilizing this feature.

Lets investigate how I would test the changes from an external project utilizing git subrepo.
I would first `git subrepo pull` If I am feeling lucky I may pass in `--all` then run my tests. Because git makes it easy to roll back any changes I don't feel there is a need to create a worktree for this work. If I am interrupted for a bugfix before I have completed my testing I would then create a worktree to address that.
My co-worker on the other hand does everything in a worktree, and as long as you create a branch for your work git subrepo works very well in a worktree.

In summary I believe git subrepo works very well with the explanation you supplied for issue number 3.

**Comment by andry81** [comment]
> 1. subtree merge with sparse checkout (partial tree checkout).

>    to contain a list of files

That as such is not required. Is enough to takeout only directories. It is definitely not as difficult. At least I could make it in the [pyxvcs](https://github.com/andry81/pyxvcs) project in python. But it is in dormant state currently.

> 2. Single list for all externals

> Why is this a necessary feature?

To merge and maintain it as a standalone, manually add and remove references to externals, change hashes and branches and so on.

> 3. external link to track head or branch tip

> This has caused me a lot of problems on projects that were utilizing this feature.

You just take out the last state of externals by checkout. It does not mean to immediately merge it. Yes, this might be a problem in the git, because the checkout is needed without a merge. But that is not meant it is not required.

> In summary I believe git subrepo works very well with the explanation you supplied for issue number 3.

If so, then ok.

**Comment by admorgan** [comment]
2. single list for all externals
This is already possible to maintain it as a standalone, manually add and remove references to externals in git subrepo by modifying the .gitrepo file(s).

**Comment by frank-dspeed** [comment]
@andry81 @admorgan i have good news for you o overwork git and git flows at present on a total new level on the root.

Let me explain what you can expect to get really fresh next year.

I Invented a new File Storage System based on btrfs why btrfs would go to deep importent here is btrfs is supported in all environments even if it is only via loopback file in the most bad case but support is there.

i also implemented parts of btrfs in userland and there comes the real magic when working with git and large repos 

i took ideas from git annex dag p2p trees and many other stuff like block deduplication algos and joined all my knowleg together.

## What has that to do with git
lets jump into git details deep as you git lovers a git commit is in fact created out of hashes its a block chain so if you filesystem has already all needed hashes to do commits you can do virtual and real time commits on the fly without handling additional real content. 

This Enables based on Real File Content Virtual Git Something like Microsoft Git FS but on steorids. 

most best is it is 1:1 svn compatible it would be just a other view of the File System no real modification is needed.

this again enables something that i guess will speed up development world wide as we can then jump into the next step 

building virtual dag trees (IPFS) Torrent and directly p2p share git repos with autosync 

there are tons of futures but all are enabled not via  changing existing tooling they are enabled via using the existing tooling correct in a consistent way on the new File System 

as the tool like git svn or anything else is only aware of the current folder view on a filesystem it can not do much and needs to store additional meta in that folder but when the file system is aware of the whole view it can supply that meta data on the fly hope that makes sense.

i am 100% sure you will love it i invested much brain power to solve developer needs at the root cause and not fix the sympthoms.

**Comment by andry81** [comment]
@frank-dspeed I am not sure how a file system is related to all said here, but anyway thank you for details and explanation. You definitely must put somehow a link to this issue into readme file.

**Comment by frank-dspeed** [comment]
@andry81 it is more simple then you think at the end this is simply never needed by anyone while there are people favoring this kind of workflow at the end. The Resulting Worktree is the same no matter what method you choose.

And a filesystem has everything to do with it as git is a filesystem in userland it adds file meta data in a .git folder or creates even only a meta data folder (bare repo) that tracks file meta data (file attributes, size, content hash, commit messages,) This stuff is also part of any filesystem at its core.

**Comment by andry81** [comment]
> And a filesystem has everything to do with it as git is a filesystem in userland it adds file meta data in a .git folder or creates even only a meta data folder (bare repo) that tracks file meta data (file attributes, size, content hash, commit messages,) This stuff is also part of any filesystem at its core.
>
> git uses the meta data to do rollbacks so it can be seen as snapshot every filesystem that offers version snapshots is a git replacement.

Yes, this is brought by the .git subdirectory which is ok. In case of a file system there would be a hidden not always copyable or transferable state (at least by a simple file copy to a file system where no such a place for the specific metadata), which is kind of dangerous. Or you does mean something different under phrase "a file system".

**Comment by frank-dspeed** [comment]
@andry81 nope it means a real real filesystem a not even block based one :) a Real Low Level File System the cool thing about it is that i use the same infos to show the same stuff in diffrent combinations so lets say more simple

data can have diffrent views you can watch the same folder as git as svn as p2p shared dag or anything else you can do with files or folders but i do not store that i create such data realtime.

**Comment by andry81** [comment]
> data can have diffrent views you can watch the same folder as git as svn as p2p shared dag or anything else you can do with files or folders but i do not store that i create such data realtime.

I don't think it is good idea. All these "subfolders" can be easely in desync state to each other which does break one or another "subfolder" because processed by different applications. Better just put `.svn`, `.git` or `.whatever` as simple metadata true subdirectory to at least not hide such behaviour.

**Comment by frank-dspeed** [comment]
@andry81 it is the filesystem if something is not there it simply does not exist it can not get desync it is a filesystem this is low level 

Linux has a wealth of filesystems from which to choose, but we are facing a number of challenges with scaling to the large storage subsystems that are becoming common in today's data centers. Filesystems need to scale in their ability to address and manage large storage, and also in their ability to detect, repair and tolerate errors in the data stored on disk.

Major Features Currently Implemented
Extent based file storage
2^64 byte == 16 EiB maximum file size (practical limit is 8 EiB due to Linux VFS)
[Space-efficient packing of small files](https://btrfs.readthedocs.io/en/latest/Inline-files.html)
Space-efficient indexed directories
Dynamic inode allocation
Writable snapshots, read-only snapshots
[Subvolumes](https://btrfs.readthedocs.io/en/latest/Subvolumes.html) (separate internal filesystem roots)
[Checksums on data and metadata](https://btrfs.readthedocs.io/en/latest/Checksumming.html) (crc32c, xxhash, sha256, blake2b)
[Compression](https://btrfs.readthedocs.io/en/latest/Compression.html) (ZLIB, LZO, ZSTD), heuristics
Integrated [multiple device support](https://btrfs.wiki.kernel.org/index.php/Multiple_Device_Support)
File Striping
File Mirroring
File Striping+Mirroring
Single and Dual Parity implementations (experimental, not production-ready)
SSD (flash storage) awareness
[TRIM/Discard](https://btrfs.readthedocs.io/en/latest/Trim.html) for reporting free blocks for reuse
Optimizations (e.g. avoiding unnecessary seek optimizations, sending writes in clusters, even if they are from unrelated files. This results in larger write operations and faster write throughput)
Background [scrub](https://btrfs.readthedocs.io/en/latest/Scrub.html) process for finding and repairing errors of files with redundant copies
Online filesystem [defragmentation](https://btrfs.readthedocs.io/en/latest/Defragmentation.html)
[Offline filesystem check](https://btrfs.readthedocs.io/en/latest/btrfs-check.html)
In-place [conversion](https://btrfs.wiki.kernel.org/index.php/Conversion_from_Ext3) of existing ext2/3/4 and reiserfs file systems
[Seeding devices](https://btrfs.readthedocs.io/en/latest/Seeding-device.html). Create a (readonly) filesystem that acts as a template to seed other Btrfs filesystems. The original filesystem and devices are included as a readonly starting point for the new filesystem. Using copy on write, all modifications are stored on different devices; the original is unchanged.
Subvolume-aware [quota support](https://btrfs.readthedocs.io/en/latest/Qgroups.html)
[Send/receive](https://btrfs.readthedocs.io/en/latest/Send-receive.html) of subvolume changes, efficient incremental filesystem mirroring and [backup](https://btrfs.wiki.kernel.org/index.php/Incremental_Backup)
Batch, or out-of-band [deduplication](https://btrfs.wiki.kernel.org/index.php/Deduplication) (happens after writes, not during)
[Swapfile](https://btrfs.readthedocs.io/en/latest/Swapfile.html) support
[Tree-checker](https://btrfs.readthedocs.io/en/latest/Tree-checker.html), post-read and pre-write metadata verification
[Zoned](https://btrfs.wiki.kernel.org/index.php/Zoned) mode support (SMR/ZBC/ZNS friendly allocation)
fsverity integration
Features by kernel version
[features by kernel version](https://btrfs.wiki.kernel.org/index.php/Changelog#By_feature)
Features Currently in Development or Planned for Future Implementation
The file/directory -level encryption support (fscrypt)
DAX/persistent memory support

**Comment by frank-dspeed** [comment]
https://nerdvittles.com/the-one-minute-installer-deploying-asterisk-on-the-internet/
